### PR TITLE
Atmega328p adc support

### PR DESCRIPTION
**Pull request by mali**
This is a basic adc support for arduino atmega328p based boards.
* Prescaler is fixed to /128 (125kHz)
* Vref is fixed to 5V

A simple test program I've used to test with a potentiometer [here](https://github.com/mali/RIOT/blob/arduino-tests/arduino_tests/adc/main.c)

Comments needed :-)

### CODE REVIEW

File: cpu/atmega328p/periph/adc.c [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+{
+
+    /* check if the line is valid */
+    if (line >= ADC_NUMOF) {
+        return -1;
+    }
+
+    // Is it needed ...?
+//    DDRD = 0x00;  // configure PORTA as input
+//    DIDR0 = 0x00;           //Digital input disabled on all ADC ports
```
Comment (mali) [inline cpu/atmega328p/periph/adc.c]: - Improved adc by setting DDRD & DIDR0 as pointed by @plushvoxel 
- replaced REG |= _BV(REGX) avr style by REG |= (1 << REGX) notation which is widely used in RIOT.
- added analogRead() arduino sketches support in sys/arduino

**Comment by mali** [comment]
maybe @kYc0o or @aabadie can have a look too ?

File: sys/arduino/base.cpp [hunk 2]
```diff
@@ -58,3 +59,9 @@ void delay(int msec)
 {
     xtimer_usleep(1000 * msec);
 }
+
+int analogRead(int pin)
+{
+    adc_init(pin);
```
Comment (haukepetersen) [inline sys/arduino/base.cpp]: AFAIK there needs to be a conversion from the Arduino pin number to the actual RIOT specific `adc_t` ADC line here...

File: boards/arduino-common/include/periph_conf.h [hunk 3]
```diff
@@ -83,6 +83,21 @@ extern "C" {
...
 /** @} */
 
+/**
+ * @brief ADC configuration
+ *
+ * @{
+ */
+#define ADC_NUMOF       (6)
+    /* ADC Channels */
+#define ADC_CHAN_0           0 // Maybe ADC_PIN_x is better ?
```
Comment (haukepetersen) [inline boards/arduino-common/include/periph_conf.h]: > Maybe ADC_PIN_x is better ?

yes :-)

Comment (mali) [inline boards/arduino-common/include/periph_conf.h]: Done

File: sys/arduino/base.cpp (hunk 2 above)
Comment (mali) [inline sys/arduino/base.cpp]: Done

**Comment by mali** [comment]
Oops, I've made a force-push which brokes review history, sorry for that ...

**Comment by kYc0o** [comment]
Is this still WIP?

**Comment by mali** [comment]
I was waiting the 2017.01 Release to remove the WIP tag :-)

File: tests/driver_mq3/main.c [hunk 4]
```diff
@@ -55,7 +55,7 @@ int main(void)
 
         printf("RAW: %4i, per mille: %1i.%03i\n", raw, alc_a, alc_b);
 
-        xtimer_usleep(500 * 1000);
+        xtimer_usleep((uint32_t)500 * 1000);
     }
     return 0;
 }
```
Comment (kYc0o) [inline tests/driver_mq3/main.c]: I think this change should be in a separated PR. BTW, you can use the constants on `timex.h`

**Comment by kYc0o** [comment]
Do you think it would be hard to add also the mega2560 and mega1281 cpus? Maybe adc.c can also be in arduino_common code?

**Comment by mali** [comment]
@kYc0o I can have a look for mega2560 and mega1281 but I don't have the hardware to test it.

**Comment by kYc0o** [comment]
I have the hardware to test, I'd like just to know how much different it can be, just to avoid another PR factorising the code to atmega_common.

**Comment by mali** [comment]
Seems quite similar, except 2560 has 16 adc while 1281 and 328p habe only 8. 
Will push a new PR in a few days.

**Comment by kYc0o** [comment]
@mali Thanks!

File: tests/driver_mq3/main.c (hunk 4 above)
Comment (mali) [inline tests/driver_mq3/main.c]: PR send on #6581

**Comment by mali** [comment]
I close this one, a better solution is proposed with PR #6616

**Pull request closed without merging**